crates/vm/levm/src/precompiles.rs

```diff
@@ -1007,22 +1011,90 @@ pub fn bls12_g1add(
+            || calldata[128..144] != [0; 16]
+            || calldata[192..208] != [0; 16]
+        {
+            return Err(PrecompileError::ParsingInputError.into());
+        }
 
-    if result_of_addition.is_identity().into() {
-        return Ok(Bytes::copy_from_slice(&G1_POINT_AT_INFINITY));
-    }
+        fn parse_g1_point(
+            x_data: &[u8],
+            y_data: &[u8],
+        ) -> Result<(BLS12381FieldElement, BLS12381FieldElement), PrecompileError> {
+            let modulus_bytes = BLS12381FieldModulus::MODULUS.limbs.map(|x| x.to_be_bytes());
+            let modulus_bytes = modulus_bytes.as_flattened();
+            if (x_data.len() == modulus_bytes.len() && x_data >= modulus_bytes)
+                || (y_data.len() == modulus_bytes.len() && y_data >= modulus_bytes)
+            {
+                return Err(PrecompileError::ParsingInputError);
+            }
 
-    let result_bytes = G1Affine::from(result_of_addition).to_uncompressed();
+            let x = BLS12381FieldElement::from_bytes_be(x_data).unwrap();
+            let y = BLS12381FieldElement::from_bytes_be(y_data).unwrap();
 
-    let mut padded_result = Vec::with_capacity(128);
-    add_padded_coordinate(&mut padded_result, &result_bytes[0..48]);
-    add_padded_coordinate(&mut padded_result, &result_bytes[48..96]);
+            if (x == BLS12381FieldElement::zero() && y == BLS12381FieldElement::zero())
+                || BLS12381Curve::defining_equation(&x, &y) == BLS12381FieldElement::zero()
+            {
+                Ok((x, y))
+            } else {
+                Err(PrecompileError::BLS12381G1PointNotInCurve)
+            }
+        }
 
-    Ok(Bytes::from(padded_result))
+        let p0 = parse_g1_point(&calldata[16..64], &calldata[80..128])?;
+        let p1 = parse_g1_point(&calldata[144..192], &calldata[208..256])?;
+
+        let p2 = if p0.0 == BLS12381FieldElement::zero() && p0.1 == BLS12381FieldElement::zero() {
+            p1
+        } else if p1.0 == BLS12381FieldElement::zero() && p1.1 == BLS12381FieldElement::zero() {
+            p0
+        } else if p0.0 == p1.0 && &p0.1 + &p1.1 == BLS12381FieldElement::zero() {
+            (BLS12381FieldElement::zero(), BLS12381FieldElement::zero())
+        } else if p0 == p1 {
+            let x_squared = p0.0.square();
+            let s = ((&x_squared + &x_squared + &x_squared) + BLS12381Curve::a()) / (&p0.1 + &p0.1);
+            let x = s.square() - &p0.0 - &p0.0;
+            let y = s * (p0.0 - &x) - p0.1;
+            (x, y)
+        } else {
+            let y01 = &p0.1 - p1.1;
+            let x01 = &p0.0 - &p1.0;
+            match x01.inv() {
+                Ok(x10_inv) => {
+                    let l = y01 * x10_inv;
+
+                    let x = l.square() - &p0.0 - p1.0;
+                    let y = l * (p0.0 - &x) - p0.1;
+
+                    (x, y)
+                }
+                Err(_) => unreachable!("already handled before"),
+            }
+        };
```

Comment:
This looks like something lambdaworks should be handling.

Reply:
The whole block.

Reply:
Oh, you meant the algorithm itself. It is handled, but only in projective coordinates, since it's optimized for multiple operations, but after asking about it I was told that if I only needed to add two points it is faster to do it using affine coordinates, which are not part of the library. That's the reason it's been implemented manually.